engineering/architecture/dateandtime.md

```diff
@@ -0,0 +1,107 @@
+---
+title: Human Readable Dates and Times
+description: Spec for handling dates and times in a human friendly way.
+published: true
+date: 2022-02-14T12:35:23.307Z
+tags: 
+editor: markdown
+dateCreated: 2022-02-15T14:58:57.900Z
+---
+
+## Goals
+
+1. Users should be able to interact with dates, times, datetimes, and time durations in Mathesar in the most human-friendly way possible
+2. The UI should be extremely forgiving w.r.t. date/time input, and also display dates and times in a human-friendly way.
+3. The API should be extremely forgiving w.r.t. date/time input, but should return dates, times, and time durations in a more programmatic, i.e., consistent way than the UI.
+4. The UI and API should be consistent in the following way:  If the input string is understood by both UI and API, they should always result in the _same date/time_ being stored, regardless of the input method.
```

Comment:
> If the input string is understood by both UI and API, they should always result in the _same date/time_ being stored, regardless of the input method.

Agreed!

I want to highlight a [related conversation](https://github.com/centerofci/mathesar/discussions/1057) in which I said:

> Some situations may have edge cases which create difficulty in making the client's results consistent with the server's results.

and @kgodey [said](https://github.com/centerofci/mathesar/discussions/1057#discussioncomment-2146476):

> checking if something is an ISO datetime shouldn't be that hard

I know we're not critiquing implementation yet, but I just want to offer a thought for people to consider early-on during implementation...

If we had a Node backend, reusing the same parsing code across client and server would be trivial, and I'd be inclined to push quite a bit of validation logic to the front end, for date/time and other stuff too. However, with our front end and back end using different languages, it's not trivial. I'm wary of implementing separate parsers (or even simple validators) in python and TypeScript, because as this document demonstrates, there are edge cases to consider.

The first thing I'd do here is to look for a library which implements a parser that we can run on both the client and server. I haven't spent time to research this, but here's an [example](https://github.com/bestowinc/json-logic-rs) of a small library I've played with before (not related to dates/times) that runs the same code in Python and a browser (via wasm). If such a library already exists for parsing ISO 8601 and/or RFC 3339, then I'd be inclined to pursue it a bit. If we did decide to use a 3rd-party library, then all these specs would become subject to its implemenation details -- which is why I rais this point now, instead of during our own implemenation phase.

@mathemancer, have you looked for any 3rd-party libs that already do this parsing?

If a suitable 3rd-party library doesn't already exist, then I'm a bit torn about how I'd want to proceed. Personally, I'd be quite interested in writing one in Rust, but that's likely more than work that we want to bite off right now. I would accept an approach where we this logic in Python and TypeScript, if others are really set on it. But if everything were up to me, I'd probably begin by delegating full responsibility to the back-end for all the date/time parsing.

Reply:
> A lot of what I'm doing right now is avoiding letting python have anything to do with date and/or time parsing

Ok I'm confused then. Can you help me understand? For example, if the user imports a CSV file containing a string that we'd like to parse into a date, where would the code exist that would do that parsing?

Reply:
[Arrow](https://github.com/arrow-py/arrow), a python library for parsing DateTime is quite similar to `moment.js`, so we could consider using it if necessary. We are already using it to validate timestamp-related display options. But I am not sure if we should be parsing on the service layer as most of the python libraries would be converting it to a python based `timedelta` which have issues with the [precision](https://github.com/centerofci/mathesar/pull/1067).

Reply:
What language are you currently planning to use to write the parsing logic?

Reply:
I don't think the UI should understand date/time _while_ the user is entering it. The _only_ purpose in this case would be to perform validations, and I think validations should always be done on the server for this.

* We intend to allow the user to enter any string which may be a date or time. Eg., 'yesterday', 'an hour ago'.
* We would probably want to extend these in the future (with more such common strings or even additional languages).
* We can only do that without conflict if we did this in one place: either the server or the client and it makes most sense to do it on the server.

_After_ the user enters it, the UI will absolutely need to understand valid date/time strings. This would be possible as the server will always respond with a well-defined format as mentioned further down in the spec below. This can be used for any operation the UI needs as well as for formatting the strings based on display options.

Reply:
> Ok I didn't see this edit you made when I wrote my previous comment. Have you considered using something like [plv8](https://github.com/plv8/plv8) and writing all the parsing logic in JavaScript? I've never done this but I've heard good things about it.

I'd need to do anything custom in PL/PGSQL (or just SQL).  The reason is that we can't count on being able to install extensions.  Some users may be using managed PosgreSQL instances (e.g., Amazon's RDS) where installing extensions isn't possible.